torch_xla/core/dynamo_bridge.py

```diff
@@ -447,6 +479,27 @@ def optimized_mod(*args: tuple):
     nonlocal dumb_return_handler
     nonlocal xla_args_need_update
     nonlocal skip_checking_input_sharding_threashold
+    nonlocal input_shape_mappings
+
+    # See [Note: Dynamo real-time input-shape cache look-up] above.
+    if not torch._dynamo.config.assume_static_by_default:
+      xla_model.xla_args = args
+      arg_input_shapes = _get_arg_input_shapes(xla_model.xla_args)
+      if arg_input_shapes in input_shape_mappings:
+        (xla_args_sharding_spec, args_and_out, graph_hash,
+         arg_index_to_need_update_index, none_remover, graph_input_matcher,
+         dumb_return_handler,
+         xla_args_need_update) = input_shape_mappings[arg_input_shapes]
+      else:
+        xm.mark_step()
+        (xla_args_sharding_spec, args_and_out, graph_hash,
+         arg_index_to_need_update_index, none_remover, graph_input_matcher,
+         dumb_return_handler,
+         xla_args_need_update) = extract_graph_helper(xla_model)
+        input_shape_mappings[arg_input_shapes] = (
+            xla_args_sharding_spec, args_and_out, graph_hash,
+            arg_index_to_need_update_index, none_remover, graph_input_matcher,
+            dumb_return_handler, xla_args_need_update)
```

Comment:
I think you don't need this here

Reply:
IIUC, we actually need this here. And we actually don't need this same logic in `extract_internal` above (removed this in the newest commit). The reason is when `dynamic=True`, only `optimized_mod` is called. Other functions (including `extract_internal`) are not called.

Reply:
ok then you will run into the same old problem right?
first time
```
extract_graph_helper -> optimized_mod
```
in this case you do the compile, but you do not cache the `input_shape_mappings `

when `optimized_mod ` is called the first tiem you will need to call `extract_graph_helper` again which is wasteful.

Reply:
you should just do the caching(input_shape_mappings[arg_input_shapes] =) inside the `extract_graph_helper `

Reply:
let me fix this too..